tests/failpoints/cases/test_merge.rs

```diff
@@ -653,77 +637,220 @@ fn test_node_merge_restart_after_apply_premerge_before_apply_compact_log() {
+
+    let left_peer_3 = find_peer(&left, 3).cloned().unwrap();
+    assert_eq!(left_peer_3.get_id(), 1003);
+
+    // Prevent sched_merge_tick to propose CommitMerge
+    let schedule_merge_fp = "on_schedule_merge";
+    fail::cfg(schedule_merge_fp, "return()").unwrap();
+
+    // To minimize peers log gap for merging
+    cluster.must_put(b"k11", b"v2");
+    must_get_equal(&cluster.get_engine(2), b"k11", b"v2");
+    must_get_equal(&cluster.get_engine(3), b"k11", b"v2");
+    // Make peer 1003 can't receive PrepareMerge and RollbackMerge log
+    cluster.add_send_filter(IsolationFilterFactory::new(3));
+
+    cluster.try_merge(left.get_id(), right.get_id());
+
+    // Change right region's epoch to make this merge failed
+    cluster.must_split(&right, b"k8");
+    fail::remove(schedule_merge_fp);
+    // Wait for left region to rollback merge
+    sleep_ms(200);
+    // Prevent the `PrepareMerge` and `RollbackMerge` log sending to apply fsm after
+    // cleaning send filter. Since this method is just to check `RollbackMerge`,
+    // the `PrepareMerge` may escape, but it makes the best effort.
+    let before_send_rollback_merge_1003_fp = "before_send_rollback_merge_1003";
+    fail::cfg(before_send_rollback_merge_1003_fp, "pause").unwrap();
+    cluster.clear_send_filters();
+
+    right = pd_client.get_region(b"k5").unwrap();
+    let right_peer_1 = find_peer(&right, 1).cloned().unwrap();
+    cluster.must_transfer_leader(right.get_id(), right_peer_1);
+    // Add some data for checking data integrity check at a later time
+    for i in 0..5 {
+        cluster.must_put(format!("k2{}", i).as_bytes(), b"v3");
+    }
+    // Do a really succeed merge
+    pd_client.must_merge(left.get_id(), right.get_id());
+    // Wait right region to send CatchUpLogs to left region.
+    sleep_ms(100);
+    // After executing CatchUpLogs in source peer fsm, the committed log will send
+    // to apply fsm in the end of this batch. So even the first `on_ready_prepare_merge`
+    // is executed after CatchUplogs, the latter committed logs is still sent to apply fsm
+    // if CatchUpLogs and `on_ready_prepare_merge` is in different batch.
+    //
+    // In this case, the data is complete because the wrong up-to-date msg from the
+    // first `on_ready_prepare_merge` is sent after all committed log.
+    // Sleep a while to wait apply fsm to send `on_ready_prepare_merge` to peer fsm.
+    let after_send_to_apply_1003_fp = "after_send_to_apply_1003";
+    fail::cfg(after_send_to_apply_1003_fp, "sleep(300)").unwrap();
+
+    fail::remove(before_send_rollback_merge_1003_fp);
+    // Wait `after_send_to_apply_1003` timeout
+    sleep_ms(300);
+    fail::remove(after_send_to_apply_1003_fp);
+    // Check the data integrity
+    for i in 0..5 {
+        must_get_equal(&cluster.get_engine(3), format!("k2{}", i).as_bytes(), b"v3");
+    }
+}
+
+// Test if the source region is destroyed correctly in merging when meeting transfer leader.
```

Comment:
How about
```
/// Tests whether the source peer is destroyed correctly when transferring leader during committing merge.
///
/// In the previous merge flow, target peer deletes meta of source peer without marking it as pending remove.
/// If source peer becomes leader at the same time, it will panic due to corrupted meta.
```